File: 程序员面试金典/0106 - 字符串压缩/Solution.js
/**
 * 面试题 01.06. 字符串压缩 (https://leetcode-cn.com/problems/compress-string-lcci/)
 * Difficulty: Easy
 *
 * 执行用时: 76 ms, 在所有 JavaScript 提交中击败了 61.00% 的用户
 * 内存消耗: 38.4 MB, 在所有 JavaScript 提交中击败了 100.00% 的用户
 * @param {string} S
 * @return {string}
 */
var compressString = function (S) {
  let sLength = S.length;
  if (sLength <= 2) {
    return S;
  }

  let compressString = "";
  let firstPointer = 0;
  let secondPointer = 1;

  let firstChar = S[firstPointer];
  let secondChar;

  while (secondPointer < sLength) {
    if (compressString.length >= sLength) {
      return S;
    }
    secondChar = S[secondPointer];

    if (firstChar !== secondChar) {
      compressString += `${firstChar}${secondPointer - firstPointer}`;
      firstPointer = secondPointer;
      firstChar = secondChar;
      secondPointer++;
    } else {
      secondPointer++;
    }
  }

  compressString += `${firstChar}${secondPointer - firstPointer}`;

  return compressString.length >= sLength ? S : compressString;
};

File: 程序员面试金典/0106 - 字符串压缩/Solution.cs
/// 面试题 01.06. 字符串压缩 (https://leetcode-cn.com/problems/compress-string-lcci/)
/// Difficulty: Easy
/// 
/// 执行用时: 384 ms, 在所有 C# 提交中击败了 40.57% 的用户
/// 内存消耗: 45.3 MB, 在所有 C# 提交中击败了 100.00% 的用户

public class Solution
{
    public string CompressString(string S)
    {
        int sLength = S.Length;
        if (sLength <= 2) { return S; }

        string compressedStr = "";
        int firstPointer = 0;
        int secondPointer = 1;

        char firstChar = S[firstPointer];
        char secondChar;

        while (secondPointer < sLength)
        {
            if (compressedStr.Length >= sLength) { return S; }
            secondChar = S[secondPointer];

            if (secondChar.Equals(firstChar))
            {
                secondPointer++;
            }
            else
            {
                compressedStr += $"{firstChar}{secondPointer - firstPointer}";
                firstPointer = secondPointer;
                firstChar = S[firstPointer];

                secondPointer++;
            }
        }

        compressedStr += $"{S[firstPointer]}{secondPointer - firstPointer}";

        return compressedStr.Length < sLength ? compressedStr : S;
    }
}